File: examples/zig/tic-tac-toe/TicTacToe.runar.zig
const runar = @import("runar");

pub const TicTacToe = struct {
    pub const Contract = runar.StatefulSmartContract;

    playerX: runar.PubKey,
    betAmount: i64,
    p2pkhPrefix: runar.ByteString,
    p2pkhSuffix: runar.ByteString,
    playerO: runar.PubKey = "000000000000000000000000000000000000000000000000000000000000000000",
    c0: i64 = 0,
    c1: i64 = 0,
    c2: i64 = 0,
    c3: i64 = 0,
    c4: i64 = 0,
    c5: i64 = 0,
    c6: i64 = 0,
    c7: i64 = 0,
    c8: i64 = 0,
    turn: i64 = 0,
    status: i64 = 0,

    pub fn init(playerX: runar.PubKey, betAmount: i64) TicTacToe {
        return .{
            .playerX = playerX,
            .betAmount = betAmount,
            .p2pkhPrefix = "1976a914",
            .p2pkhSuffix = "88ac",
        };
    }

    pub fn join(self: *TicTacToe, opponentPK: runar.PubKey, sig: runar.Sig) void {
        runar.assert(self.status == 0);
        runar.assert(runar.checkSig(sig, opponentPK));
        self.playerO = opponentPK;
        self.status = 1;
        self.turn = 1;
    }

    pub fn move(self: *TicTacToe, position: i64, player: runar.PubKey, sig: runar.Sig) void {
        runar.assert(self.status == 1);
        runar.assert(runar.checkSig(sig, player));
        self.assertCorrectPlayer(player);
        self.placeMove(position);

        if (self.turn == 1) {
            self.turn = 2;
        } else {
            self.turn = 1;
        }
    }

    pub fn moveAndWin(
        self: *const TicTacToe,
        position: i64,
        player: runar.PubKey,
        sig: runar.Sig,
        changePKH: runar.ByteString,
        changeAmount: i64,
    ) void {
        runar.assert(self.status == 1);
        runar.assert(runar.checkSig(sig, player));
        self.assertCorrectPlayer(player);
        self.assertCellEmpty(position);
        runar.assert(self.checkWinAfterMove(position, self.turn));

        const payout = runar.cat(
            runar.cat(runar.num2bin(self.betAmount * 2, 8), self.p2pkhPrefix),
            runar.cat(runar.hash160(player), self.p2pkhSuffix),
        );
        if (changeAmount > 0) {
            const change = runar.cat(
                runar.cat(runar.num2bin(changeAmount, 8), self.p2pkhPrefix),
                runar.cat(changePKH, self.p2pkhSuffix),
            );
            runar.assert(runar.hash256(runar.cat(payout, change)) == runar.extractOutputHash(self.txPreimage));
        } else {
            runar.assert(runar.hash256(payout) == runar.extractOutputHash(self.txPreimage));
        }
    }

    pub fn moveAndTie(
        self: *const TicTacToe,
        position: i64,
        player: runar.PubKey,
        sig: runar.Sig,
        changePKH: runar.ByteString,
        changeAmount: i64,
    ) void {
        runar.assert(self.status == 1);
        runar.assert(runar.checkSig(sig, player));
        self.assertCorrectPlayer(player);
        self.assertCellEmpty(position);
        runar.assert(self.countOccupied() == 8);
        runar.assert(!self.checkWinAfterMove(position, self.turn));

        const out1 = runar.cat(
            runar.cat(runar.num2bin(self.betAmount, 8), self.p2pkhPrefix),
            runar.cat(runar.hash160(self.playerX), self.p2pkhSuffix),
        );
        const out2 = runar.cat(
            runar.cat(runar.num2bin(self.betAmount, 8), self.p2pkhPrefix),
            runar.cat(runar.hash160(self.playerO), self.p2pkhSuffix),
        );
        if (changeAmount > 0) {
            const change = runar.cat(
                runar.cat(runar.num2bin(changeAmount, 8), self.p2pkhPrefix),
                runar.cat(changePKH, self.p2pkhSuffix),
            );
            runar.assert(runar.hash256(runar.cat(runar.cat(out1, out2), change)) == runar.extractOutputHash(self.txPreimage));
        } else {
            runar.assert(runar.hash256(runar.cat(out1, out2)) == runar.extractOutputHash(self.txPreimage));
        }
    }

    pub fn cancelBeforeJoin(
        self: *const TicTacToe,
        sig: runar.Sig,
        changePKH: runar.ByteString,
        changeAmount: i64,
    ) void {
        runar.assert(self.status == 0);
        runar.assert(runar.checkSig(sig, self.playerX));

        const payout = runar.cat(
            runar.cat(runar.num2bin(self.betAmount, 8), self.p2pkhPrefix),
            runar.cat(runar.hash160(self.playerX), self.p2pkhSuffix),
        );
        if (changeAmount > 0) {
            const change = runar.cat(
                runar.cat(runar.num2bin(changeAmount, 8), self.p2pkhPrefix),
                runar.cat(changePKH, self.p2pkhSuffix),
            );
            runar.assert(runar.hash256(runar.cat(payout, change)) == runar.extractOutputHash(self.txPreimage));
        } else {
            runar.assert(runar.hash256(payout) == runar.extractOutputHash(self.txPreimage));
        }
    }

    pub fn cancel(
        self: *const TicTacToe,
        sigX: runar.Sig,
        sigO: runar.Sig,
        changePKH: runar.ByteString,
        changeAmount: i64,
    ) void {
        const out1 = runar.cat(
            runar.cat(runar.num2bin(self.betAmount, 8), self.p2pkhPrefix),
            runar.cat(runar.hash160(self.playerX), self.p2pkhSuffix),
        );
        const out2 = runar.cat(
            runar.cat(runar.num2bin(self.betAmount, 8), self.p2pkhPrefix),
            runar.cat(runar.hash160(self.playerO), self.p2pkhSuffix),
        );
        if (changeAmount > 0) {
            const change = runar.cat(
                runar.cat(runar.num2bin(changeAmount, 8), self.p2pkhPrefix),
                runar.cat(changePKH, self.p2pkhSuffix),
            );
            runar.assert(runar.hash256(runar.cat(runar.cat(out1, out2), change)) == runar.extractOutputHash(self.txPreimage));
        } else {
            runar.assert(runar.hash256(runar.cat(out1, out2)) == runar.extractOutputHash(self.txPreimage));
        }

        runar.assert(runar.checkSig(sigX, self.playerX));
        runar.assert(runar.checkSig(sigO, self.playerO));
    }

    fn assertCorrectPlayer(self: *const TicTacToe, player: runar.PubKey) void {
        if (self.turn == 1) {
            runar.assert(player == self.playerX);
        } else {
            runar.assert(player == self.playerO);
        }
    }

    fn assertCellEmpty(self: *const TicTacToe, position: i64) void {
        if (position == 0) {
            runar.assert(self.c0 == 0);
        } else if (position == 1) {
            runar.assert(self.c1 == 0);
        } else if (position == 2) {
            runar.assert(self.c2 == 0);
        } else if (position == 3) {
            runar.assert(self.c3 == 0);
        } else if (position == 4) {
            runar.assert(self.c4 == 0);
        } else if (position == 5) {
            runar.assert(self.c5 == 0);
        } else if (position == 6) {
            runar.assert(self.c6 == 0);
        } else if (position == 7) {
            runar.assert(self.c7 == 0);
        } else if (position == 8) {
            runar.assert(self.c8 == 0);
        } else {
            runar.assert(false);
        }
    }

    fn placeMove(self: *TicTacToe, position: i64) void {
        self.assertCellEmpty(position);
        if (position == 0) {
            self.c0 = self.turn;
        } else if (position == 1) {
            self.c1 = self.turn;
        } else if (position == 2) {
            self.c2 = self.turn;
        } else if (position == 3) {
            self.c3 = self.turn;
        } else if (position == 4) {
            self.c4 = self.turn;
        } else if (position == 5) {
            self.c5 = self.turn;
        } else if (position == 6) {
            self.c6 = self.turn;
        } else if (position == 7) {
            self.c7 = self.turn;
        } else if (position == 8) {
            self.c8 = self.turn;
        } else {
            runar.assert(false);
        }
    }

    fn getCellOrOverride(self: *const TicTacToe, cellIndex: i64, overridePos: i64, overrideVal: i64) i64 {
        if (cellIndex == overridePos) {
            return overrideVal;
        }
        if (cellIndex == 0) {
            return self.c0;
        }
        if (cellIndex == 1) {
            return self.c1;
        }
        if (cellIndex == 2) {
            return self.c2;
        }
        if (cellIndex == 3) {
            return self.c3;
        }
        if (cellIndex == 4) {
            return self.c4;
        }
        if (cellIndex == 5) {
            return self.c5;
        }
        if (cellIndex == 6) {
            return self.c6;
        }
        if (cellIndex == 7) {
            return self.c7;
        }
        return self.c8;
    }

    fn checkWinAfterMove(self: *const TicTacToe, position: i64, player: i64) bool {
        const c0 = self.getCellOrOverride(0, position, player);
        const c1 = self.getCellOrOverride(1, position, player);
        const c2 = self.getCellOrOverride(2, position, player);
        const c3 = self.getCellOrOverride(3, position, player);
        const c4 = self.getCellOrOverride(4, position, player);
        const c5 = self.getCellOrOverride(5, position, player);
        const c6 = self.getCellOrOverride(6, position, player);
        const c7 = self.getCellOrOverride(7, position, player);
        const c8 = self.getCellOrOverride(8, position, player);

        return
            (c0 == player and c1 == player and c2 == player) or
            (c3 == player and c4 == player and c5 == player) or
            (c6 == player and c7 == player and c8 == player) or
            (c0 == player and c3 == player and c6 == player) or
            (c1 == player and c4 == player and c7 == player) or
            (c2 == player and c5 == player and c8 == player) or
            (c0 == player and c4 == player and c8 == player) or
            (c2 == player and c4 == player and c6 == player);
    }

    fn countOccupied(self: *const TicTacToe) i64 {
        var count: i64 = 0;
        if (self.c0 != 0) {
            count += 1;
        }
        if (self.c1 != 0) {
            count += 1;
        }
        if (self.c2 != 0) {
            count += 1;
        }
        if (self.c3 != 0) {
            count += 1;
        }
        if (self.c4 != 0) {
            count += 1;
        }
        if (self.c5 != 0) {
            count += 1;
        }
        if (self.c6 != 0) {
            count += 1;
        }
        if (self.c7 != 0) {
            count += 1;
        }
        if (self.c8 != 0) {
            count += 1;
        }
        return count;
    }
};

File: examples/go/tic-tac-toe/TicTacToe.runar.go
package contract

import runar "github.com/icellan/runar/packages/runar-go"

// TicTacToe is an on-chain Tic-Tac-Toe contract.
//
// Two players compete on a 3x3 board. Each move is an on-chain transaction.
// The contract holds both players' bets and enforces correct game rules
// entirely in Bitcoin Script.
//
// Board encoding:
// Since Rúnar has no arrays, the 3x3 board uses 9 individual bigint fields
// (C0-C8). Values: 0=empty, 1=X, 2=O.
//
// Lifecycle:
//  1. Player X deploys the contract with their bet amount.
//  2. Player O calls Join to enter the game, adding their bet.
//  3. Players alternate calling Move (non-terminal) or
//     MoveAndWin / MoveAndTie (terminal).
//  4. Either player can propose Cancel (requires both signatures).
//
// Method types:
//   - State-mutating: Join, Move — produce a continuation UTXO.
//   - Non-mutating terminal: MoveAndWin, MoveAndTie, Cancel — spend
//     the UTXO and enforce payout outputs via ExtractOutputHash.
type TicTacToe struct {
	runar.StatefulSmartContract
	PlayerX     runar.PubKey      `runar:"readonly"`
	BetAmount   runar.Bigint      `runar:"readonly"`
	P2pkhPrefix runar.ByteString  `runar:"readonly"`
	P2pkhSuffix runar.ByteString  `runar:"readonly"`
	PlayerO     runar.PubKey
	C0          runar.Bigint
	C1          runar.Bigint
	C2          runar.Bigint
	C3          runar.Bigint
	C4          runar.Bigint
	C5          runar.Bigint
	C6          runar.Bigint
	C7          runar.Bigint
	C8          runar.Bigint
	Turn        runar.Bigint
	Status      runar.Bigint
}

func (c *TicTacToe) init() {
	c.P2pkhPrefix = "1976a914"
	c.P2pkhSuffix = "88ac"
	c.PlayerO = "000000000000000000000000000000000000000000000000000000000000000000"
	c.C0 = 0
	c.C1 = 0
	c.C2 = 0
	c.C3 = 0
	c.C4 = 0
	c.C5 = 0
	c.C6 = 0
	c.C7 = 0
	c.C8 = 0
	c.Turn = 0
	c.Status = 0
}

// Join allows Player O to join the game.
// State-mutating: produces continuation UTXO with doubled bet.
func (c *TicTacToe) Join(opponentPK runar.PubKey, sig runar.Sig) {
	runar.Assert(c.Status == 0)
	runar.Assert(runar.CheckSig(sig, opponentPK))
	c.PlayerO = opponentPK
	c.Status = 1
	c.Turn = 1
}

// Move makes a non-terminal move. Updates board and flips turn.
// State-mutating: produces continuation UTXO.
// Caller provides their pubkey; contract verifies it matches the expected turn.
func (c *TicTacToe) Move(position runar.Bigint, player runar.PubKey, sig runar.Sig) {
	runar.Assert(c.Status == 1)
	runar.Assert(runar.CheckSig(sig, player))
	c.assertCorrectPlayer(player)
	c.placeMove(position)
	if c.Turn == 1 {
		c.Turn = 2
	} else {
		c.Turn = 1
	}
}

// MoveAndWin makes a winning move. Non-mutating terminal method.
// Enforces winner-gets-all payout via ExtractOutputHash.
// Supports optional change output for fee funding.
func (c *TicTacToe) MoveAndWin(position runar.Bigint, player runar.PubKey, sig runar.Sig, changePKH runar.ByteString, changeAmount runar.Bigint) {
	runar.Assert(c.Status == 1)
	runar.Assert(runar.CheckSig(sig, player))
	c.assertCorrectPlayer(player)
	c.assertCellEmpty(position)
	runar.Assert(c.checkWinAfterMove(position, c.Turn))

	totalPayout := c.BetAmount * 2
	payout := runar.Cat(runar.Cat(runar.Num2Bin(totalPayout, 8), c.P2pkhPrefix), runar.Cat(runar.Hash160(player), c.P2pkhSuffix))
	if changeAmount > 0 {
		change := runar.Cat(runar.Cat(runar.Num2Bin(changeAmount, 8), c.P2pkhPrefix), runar.Cat(changePKH, c.P2pkhSuffix))
		runar.Assert(runar.Hash256(runar.Cat(payout, change)) == runar.ExtractOutputHash(c.TxPreimage))
	} else {
		runar.Assert(runar.Hash256(payout) == runar.ExtractOutputHash(c.TxPreimage))
	}
}

// MoveAndTie makes a move that fills the board (tie). Non-mutating terminal method.
// Enforces equal split payout via ExtractOutputHash.
// Supports optional change output for fee funding.
func (c *TicTacToe) MoveAndTie(position runar.Bigint, player runar.PubKey, sig runar.Sig, changePKH runar.ByteString, changeAmount runar.Bigint) {
	runar.Assert(c.Status == 1)
	runar.Assert(runar.CheckSig(sig, player))
	c.assertCorrectPlayer(player)
	c.assertCellEmpty(position)
	runar.Assert(c.countOccupied() == 8)
	runar.Assert(!c.checkWinAfterMove(position, c.Turn))

	out1 := runar.Cat(runar.Cat(runar.Num2Bin(c.BetAmount, 8), c.P2pkhPrefix), runar.Cat(runar.Hash160(c.PlayerX), c.P2pkhSuffix))
	out2 := runar.Cat(runar.Cat(runar.Num2Bin(c.BetAmount, 8), c.P2pkhPrefix), runar.Cat(runar.Hash160(c.PlayerO), c.P2pkhSuffix))
	if changeAmount > 0 {
		change := runar.Cat(runar.Cat(runar.Num2Bin(changeAmount, 8), c.P2pkhPrefix), runar.Cat(changePKH, c.P2pkhSuffix))
		runar.Assert(runar.Hash256(runar.Cat(runar.Cat(out1, out2), change)) == runar.ExtractOutputHash(c.TxPreimage))
	} else {
		runar.Assert(runar.Hash256(runar.Cat(out1, out2)) == runar.ExtractOutputHash(c.TxPreimage))
	}
}

// CancelBeforeJoin lets Player X cancel before anyone joins. Non-mutating terminal method.
// Refunds the full bet to Player X.
// Supports optional change output for fee funding.
func (c *TicTacToe) CancelBeforeJoin(sig runar.Sig, changePKH runar.ByteString, changeAmount runar.Bigint) {
	runar.Assert(c.Status == 0)
	runar.Assert(runar.CheckSig(sig, c.PlayerX))
	payout := runar.Cat(runar.Cat(runar.Num2Bin(c.BetAmount, 8), c.P2pkhPrefix), runar.Cat(runar.Hash160(c.PlayerX), c.P2pkhSuffix))
	if changeAmount > 0 {
		change := runar.Cat(runar.Cat(runar.Num2Bin(changeAmount, 8), c.P2pkhPrefix), runar.Cat(changePKH, c.P2pkhSuffix))
		runar.Assert(runar.Hash256(runar.Cat(payout, change)) == runar.ExtractOutputHash(c.TxPreimage))
	} else {
		runar.Assert(runar.Hash256(payout) == runar.ExtractOutputHash(c.TxPreimage))
	}
}

// Cancel lets both players agree to cancel. Non-mutating terminal method.
// Enforces equal refund via ExtractOutputHash.
// Supports optional change output for fee funding.
func (c *TicTacToe) Cancel(sigX runar.Sig, sigO runar.Sig, changePKH runar.ByteString, changeAmount runar.Bigint) {
	out1 := runar.Cat(runar.Cat(runar.Num2Bin(c.BetAmount, 8), c.P2pkhPrefix), runar.Cat(runar.Hash160(c.PlayerX), c.P2pkhSuffix))
	out2 := runar.Cat(runar.Cat(runar.Num2Bin(c.BetAmount, 8), c.P2pkhPrefix), runar.Cat(runar.Hash160(c.PlayerO), c.P2pkhSuffix))
	if changeAmount > 0 {
		change := runar.Cat(runar.Cat(runar.Num2Bin(changeAmount, 8), c.P2pkhPrefix), runar.Cat(changePKH, c.P2pkhSuffix))
		runar.Assert(runar.Hash256(runar.Cat(runar.Cat(out1, out2), change)) == runar.ExtractOutputHash(c.TxPreimage))
	} else {
		runar.Assert(runar.Hash256(runar.Cat(out1, out2)) == runar.ExtractOutputHash(c.TxPreimage))
	}
	runar.Assert(runar.CheckSig(sigX, c.PlayerX))
	runar.Assert(runar.CheckSig(sigO, c.PlayerO))
}

// --- Private helpers ---

// assertCorrectPlayer asserts the provided player pubkey matches whoever's turn it is.
func (c *TicTacToe) assertCorrectPlayer(player runar.PubKey) {
	if c.Turn == 1 {
		runar.Assert(player == c.PlayerX)
	} else {
		runar.Assert(player == c.PlayerO)
	}
}

func (c *TicTacToe) assertCellEmpty(position runar.Bigint) {
	if position == 0 {
		runar.Assert(c.C0 == 0)
	} else if position == 1 {
		runar.Assert(c.C1 == 0)
	} else if position == 2 {
		runar.Assert(c.C2 == 0)
	} else if position == 3 {
		runar.Assert(c.C3 == 0)
	} else if position == 4 {
		runar.Assert(c.C4 == 0)
	} else if position == 5 {
		runar.Assert(c.C5 == 0)
	} else if position == 6 {
		runar.Assert(c.C6 == 0)
	} else if position == 7 {
		runar.Assert(c.C7 == 0)
	} else if position == 8 {
		runar.Assert(c.C8 == 0)
	} else {
		runar.Assert(false)
	}
}

func (c *TicTacToe) placeMove(position runar.Bigint) {
	c.assertCellEmpty(position)
	if position == 0 {
		c.C0 = c.Turn
	} else if position == 1 {
		c.C1 = c.Turn
	} else if position == 2 {
		c.C2 = c.Turn
	} else if position == 3 {
		c.C3 = c.Turn
	} else if position == 4 {
		c.C4 = c.Turn
	} else if position == 5 {
		c.C5 = c.Turn
	} else if position == 6 {
		c.C6 = c.Turn
	} else if position == 7 {
		c.C7 = c.Turn
	} else if position == 8 {
		c.C8 = c.Turn
	} else {
		runar.Assert(false)
	}
}

func (c *TicTacToe) getCellOrOverride(cellIndex runar.Bigint, overridePos runar.Bigint, overrideVal runar.Bigint) runar.Bigint {
	if cellIndex == overridePos {
		return overrideVal
	}
	if cellIndex == 0 {
		return c.C0
	} else if cellIndex == 1 {
		return c.C1
	} else if cellIndex == 2 {
		return c.C2
	} else if cellIndex == 3 {
		return c.C3
	} else if cellIndex == 4 {
		return c.C4
	} else if cellIndex == 5 {
		return c.C5
	} else if cellIndex == 6 {
		return c.C6
	} else if cellIndex == 7 {
		return c.C7
	} else {
		return c.C8
	}
}

func (c *TicTacToe) checkWinAfterMove(position runar.Bigint, player runar.Bigint) runar.Bool {
	v0 := c.getCellOrOverride(0, position, player)
	v1 := c.getCellOrOverride(1, position, player)
	v2 := c.getCellOrOverride(2, position, player)
	v3 := c.getCellOrOverride(3, position, player)
	v4 := c.getCellOrOverride(4, position, player)
	v5 := c.getCellOrOverride(5, position, player)
	v6 := c.getCellOrOverride(6, position, player)
	v7 := c.getCellOrOverride(7, position, player)
	v8 := c.getCellOrOverride(8, position, player)

	if v0 == player && v1 == player && v2 == player {
		return true
	}
	if v3 == player && v4 == player && v5 == player {
		return true
	}
	if v6 == player && v7 == player && v8 == player {
		return true
	}
	if v0 == player && v3 == player && v6 == player {
		return true
	}
	if v1 == player && v4 == player && v7 == player {
		return true
	}
	if v2 == player && v5 == player && v8 == player {
		return true
	}
	if v0 == player && v4 == player && v8 == player {
		return true
	}
	if v2 == player && v4 == player && v6 == player {
		return true
	}
	return false
}

func (c *TicTacToe) countOccupied() runar.Bigint {
	count := runar.Bigint(0)
	if c.C0 != 0 {
		count = count + 1
	}
	if c.C1 != 0 {
		count = count + 1
	}
	if c.C2 != 0 {
		count = count + 1
	}
	if c.C3 != 0 {
		count = count + 1
	}
	if c.C4 != 0 {
		count = count + 1
	}
	if c.C5 != 0 {
		count = count + 1
	}
	if c.C6 != 0 {
		count = count + 1
	}
	if c.C7 != 0 {
		count = count + 1
	}
	if c.C8 != 0 {
		count = count + 1
	}
	return count
}
